benchmarks/src/main/scala/zio/stm/TReentrantLockBenchmark.scala

```diff
@@ -141,29 +142,27 @@ class TReentrantLockBenchmark {
   @Benchmark
   @Group("ReentrantLockHighContention")
   @GroupThreads(20)
-  def reentrantLockWriteGroup3(): Unit = reentrantLockWrite()*/
+  def reentrantLockWriteGroup3(): Unit = reentrantLockWrite()
 
-  @CompilerControl(CompilerControl.Mode.DONT_INLINE)
-  def zioLockRead(): Unit = {
-    val io = for {
-      lock <- zioLock
-      _    <- lock.readLock.use(_ => doWorkM())
-    } yield ()
-
-    unsafeRun(io)
-  }
+  // used to amortize the relative cost of unsafeRun
+  // compared to benchmarked operations
+  private val calls = (0 to 10000).toList
 
   @CompilerControl(CompilerControl.Mode.DONT_INLINE)
-  def zioLockWrite(): Unit = {
-    val io = for {
-      lock <- zioLock
-      _    <- lock.writeLock.use(_ => doWorkM())
-    } yield ()
-
-    unsafeRun(io)
-  }
+  def zioLockRead(): Unit =
+    unsafeRun(
+      ZIO.foreach_(calls)(_ => zioLock.flatMap(lock => lock.acquireRead.commit *> doWorkM() *> lock.releaseRead.commit))
```

Comment:
This will measure the overhead of `foreach`. Instead, try:

```scala
def repeat(n: Int, effect: USTM[Any]): USTM[Any] = 
  if (n <= 0) effect else effect.flatMap(_ => effect)
```

Then use that in several places:

```scala
unsafeRun {
  repeat(CallCount)(lock.acquireRead.flatMap(_ => lock.releaseRead)).commit
}
```

Note: everything is important here. `zioLock` is actually a _recipe_ to create a lock, not the actual lock itself. So in some setup outside the actual benchmark, you've got to do: `zioLock = unsafeRun(ReentrantLock.make)`, where `zioLock` is a `var`. That way it will be _already created_ when the benchmark runs.

Similar changes elsewhere.

Reply:
`(lock.acquireRead.flatMap(_ => lock.releaseRead)).commit`
doesn't acquiring and releasing a read before committing it result in the lock always being in its starting state (i.e. no locks) as far as any other transactions are concerned? A lock isn't acquired until it's committed, and it's not released until it's committed again. the transaction overhead is part of the cost of using this lock, so I think it should be included in the benchmark.

Reply:
Ah, you're right, we need to commit them separately!